shell/common/renderer_context_manager.h

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_COMMON_RENDERER_CONTEXT_MANAGER_H_
+#define FLUTTER_SHELL_COMMON_RENDERER_CONTEXT_MANAGER_H_
+
+#include <functional>
+#include <memory>
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/macros.h"
+
+namespace flutter {
+
+template <typename RendererContext>
+//------------------------------------------------------------------------------
+/// Manages the renderer context.
+class RendererContextManager {
+  // Returns true if setting current context is successful.
+  // The block should be able to handle `nullptr`s as a parameter
+  using SetCurrentContextBlock =
+      std::function<bool(std::shared_ptr<RendererContext>)>;
+  // Should return a `nullptr` if there is no current context.
+  using GetCurrentContextBlock =
+      std::function<std::shared_ptr<RendererContext>()>;
+
+ public:
+  //------------------------------------------------------------------------------
+  /// Switches the renderer context to the flutter's contexts.
+  ///
+  /// In construction, it should set the current context to a flutter's context
+  /// In destruction, it should reset the current context.
+  ///
+  class RendererContextSwitch {
+   public:
+    RendererContextSwitch(RendererContextManager& manager,
+                          std::shared_ptr<RendererContext> context)
+        : manager_(manager) {
+      bool result = manager_.PushContext(context);
+      has_pushed_context_ = true;
+      result_ = result;
+    }
+
+    ~RendererContextSwitch() {
+      if (!has_pushed_context_) {
+        return;
+      }
+      manager_.PopContext();
+    }
+
+    // Returns true if the context switching was successful.
+    bool GetResult() { return result_; }
+
+   private:
+    RendererContextManager& manager_;
+    bool result_;
+    bool has_pushed_context_;
+
+    FML_DISALLOW_COPY_AND_ASSIGN(RendererContextSwitch);
+  };
+
+  RendererContextManager(std::shared_ptr<RendererContext> context,
+                         std::shared_ptr<RendererContext> resource_context,
+                         SetCurrentContextBlock set_current_block,
+                         GetCurrentContextBlock get_current_block)
+      : context_(context),
+        resource_context_(resource_context),
+        set_current_context_block_(set_current_block),
+        get_current_context_block_(get_current_block) {}
+  ~RendererContextManager() = default;
+
+  //----------------------------------------------------------------------------
+  /// @brief      Make the flutter's context as current context.
+  ///
+  /// @return     A `RendererContextSwitch` which ensures the current context is
+  /// the `context` that used in the constructor.
+  std::unique_ptr<RendererContextSwitch> MakeCurrent() {
```

Comment:
Aren't these switches tied to scoped? In that case, it isn't clear why these are unique pointers and not just lightweight value objects.